### ISSUE TITLE
[Spell] Slam! / Bloodsurge

### ISSUE THREAD

**Issue by ScoobyPwnsOnU** [WoWManiaUK/Redemption#3122]
**Links:**

http://www.wow-mania.com/armory/?item=51214

**Current Behaviour:**

I have noticed that sometimes you get the reduced global cooldown and sometimes you don't(I am not talking about not getting a reduced GCD on regular slam procs, I understand those are supposed to have a full GCD). I have noticed it seems to happen more often if you get a lucky spree of double slam procs.  So basically sometimes you get the .5 second GCD on your double slam proc, and other times you get the full 1 second GCD on both slams in your double slam proc. As far as I read it the GCD should always be reduced if you get a double slam proc.

**Expected Behaviour:**

It says when you get a double proc of slam it should reduce the global cooldown by .5 seconds for both slams, when you get a 2 charge slam proc.

**Steps to reproduce the issue:**

I have only noticed this in ICC, but I also haven't done a lot outside of ICC on the warrior since I got the 4 set bonus.
It seems to happen whether or not I have my cooldowns up, and can happen anywhere in ICC.

**Additional info:**

**Comment by wow-mania** [comment]
@Dranah can you confirm this first please?

**Comment by Dranah** [comment]
Not exactly but i found 2 other issues that needs fixing and which could explain the confusion.
1) Slam (ID 47475) does not appear to consume the bloodsurge proc (ID 46916) at all.
This means you can use multiple slams without the spell having multiple charges, which could confuse players about which proc they've got.
2) The two type of procs don't properly overwrite each other. 
It is possible to receive a second proc before you consumed the first (especially since slam never consumes it, so it actually happens a lot)
And it is possible that the two procs that follow each other aren't of the same type.
Now this is a weird situation because both the regular bloodsurge proc and the boosted one with t10 have the same name "Slam!" and same ID 46916, but they behave differently.
One is 1 charge, 5sec duration with regular 1,5sec GCd, the other is 2 charge, 10sec duration with 1 sec GCD.
Duration and charge count overwrites properly if you get two slam procs in a quick succession, however the GCD does not seem to.
Example: you get a boosted proc (2 charge, 10sec dura, 1sec gcd) and before it is consumed or expired, you get another proc, which is a regular one. The regular one overwrites it so that it becomes 1 stack and 5sec duration, but you retain the 1sec GCD.
Example2: you get a regular proc (1charge 5sec dura 1,5 gcd) and before it is consumed or expired, you get another proc, which is a boosted one. The boosted over overwrites the proc so that it becomes 2 stack and 10sec duration, but you retain the 1,5sec GCD.
This is how it can happen that you have a 2 charge proc on you but the GCD isn't reduced.

**Comment by Dranah** [comment]
Extra notes after some more tests: 
The problem may be more complex than I described at first. Something is certainly wrong with the way successive procs behave, what I can't decide if it's a visual problem or a real bug. It is possible that the boosted procs which do not have the buffed GCD are in fact regular procs that are displayed as boosted ones wrongly. Kinda impossible to tell since the two procs have same name, same ID and same icon.
Need to do a little research on how the two different effects stack on blizz, if they should fully overwrite each other or just refresh the duration of the already existing one.

**Comment by wow-mania** [comment]
Please re-test.

**Comment by Dranah** [comment]
UPDATE:
I have found some much needed extra information on how the procs should behave.

1) The regular proc and the t10 boosted proc should be clearly distinguishable by their name/combat text. Based on MMO champions , on retail the t10 boosted proc is called/labelled "Extra Slam!" as opposed to the regular "Slam!"
**How it works atm:** Here both are labelled as "Slam!" - **Need fixing!**
2) The t10 boosted Extra Slam! should never be overwritten by a regular Slam! proc, only refreshed by it. As described by sources:
_"maybe they fixed it in 3.3.2 but right now if bloodsurge refresh (normal or 4P proc), it won't bug anything, your 2 slams will still be 1sec gcd (even if it refresh at 1 charge left) and it will refresh to 10sec if it reproc while you still have 2 charges up. "_
_"Where are you getting your info that says it can be overwritten? I have NEVER had the proc overwritten.
If you get the 4pc proc, and then BS procs, all that happens is that your 2 4pc procs get reset to a 10second timer.
If you use 1 of the 4 pc procs and BS procs, you get another 2 stacks and 10 seconds, of the 4pc."_

**How it works here:** If your boosted slam proc has more than 5sec duration left, it is properly refreshed by any incoming procs (t10 or not). However, if the duration falls below 5sec, incoming regular proc becomes able to override it so that you end up losing one charge. (The gcd is properly retained though)
Basically, the problem here is that a 2 charge effect should never ever be overwritten by a 1 charge one - **need fixing**
Video made of problem:
Regular bloodsurge proc overwriting boosted one (can only happen now if the duration on boosted proc is below 5sec) https://youtu.be/kFErLoEe54o
0:06 boosted proc overwritten by boosted proc
0:13 boosted proc overwritten by regular, one stack lost

Now the issues with wrong GCD on boosted procs and some non boosted slams having shorter gcd for no reason:
I recorded these on video as well.
3) Boosted proc one charge with reduced, second charge with long GCD https://www.youtube.com/watch?v=v7cpmmWLc6c&feature=youtu.be
1:09 boosted proc with 2 charges pops
1:11 one charge consumed by slam, GCD shortened as intended
1:15 another boosted proc happens, buff is back to 2 charges and 10s duration
1:16 another boosted proc happens, buff is back to 2 charges and 10s duration
1:17 one charge consumed by slam, GCD shortened as intended
1:18 second charge consumed by slam, GCD regular long one, not intended
This version of the bug requires you to consume only one charge from the first proc and then receive a new proc quickly after.
4) Regular slams with shortened GCD after a boosted proc has expired https://youtu.be/mwzYImzJeOk
0:00 boosted proc almost expiring
0:01 one charge consumed by slam, shortened GCD as intended, buff expires after
0:02 hard casting a slam with no proc active, still shortened gcd (unintended)
0:03 hard casting another slam with no proc active,still shorter gcd (unintended)
0:05 hard casting a 3rd slam with no proc active, gcd is back to normal 
This version of the bug requires you to start chain casting slams when the proc is about the expire so that there is still a shortened gcd running when you queue up the next spell. If you simply let the buff expire and then hard cast slam, they won't have the GCD reduced.
5) No reduced GCD when getting a new double proc right after consuming the previous double proc https://www.youtube.com/watch?v=ZKbsgZhWVE0&feature=youtu.be
0:03 one charge consumed by slam from first double proc (gcd shortened as intended)
0:06 second charge consumed by slam from first double proc (gcd short as intended)
0:07 new double proc activates
0:08 one charge consumed by slam from second double proc (gcd regular, unintended)
0:11 second charge consumed by slam from second double proc (gcd regular, unintended)

Sources: https://www.mmo-champion.com/threads/700982-Fury-Warrior-T10-4PC-Clarification
https://www.mmo-champion.com/threads/694852-Your-experience-with-T10-4P-as-Fury
https://gamefaqs.gamespot.com/boards/534914-world-of-warcraft/54393421

**Comment by Keader** [comment]
Ok @Dranah , about https://github.com/WoWManiaUK/Redemption/issues/3122#issuecomment-473489304

Extra effect is called 70849 [Extra Charge!].
How it work?
It get current Slam Proc (46916) and adds 1 more charge.
That's why is always labeled as "Slam !", because well, is the same aura, extra charge just add 1 more charge of it.

And about avoid proc, why?
Right now if normal Slam proc, it just add 1 more charge of "powered one", because "extra charge" is still there.
Right now, i realy think that has no problem on this.

**Comment by Dranah** [comment]
1) _"Right now if normal Slam proc, it just add 1 more charge of "powered one", because "extra charge" is still there."_ 
This is only true if your 2charge proc has more than 5sec left on its duration. 
Otherwise (if less than 5sec duration), getting a normal slam proc on top of the powered one will simply overwrite the effect, and if you still had 2 charges remaining, you lose one charge because the new effect has only one.

2nd ) _"It get current Slam Proc (46916) and adds 1 more charge."_
Now this part is more interesting.
If you do .cast 70849 with a slam proc already active, nothing happens, so it definitely does not affect current proc.
The effect seems to work for the next incoming slam proc, so if you have 70849 active, the next proc of 46916 will have two charges and 10s duration. I suppose the mechanics are linked in a way so when you get a t10 proc, first 70849 is cast, then 46916 right after , creating a 2 charge proc together. 
There is a 3rd ID responsible for the reduced GCD, (70849 does not give boosted GCD, only an extra charge and extra duration. This ID 71072 "Slam GCD reduced" is among my auras with the proc.

Now here's what i discovered:
My tests show that an independent cast of 70849 has 10sec duration (is not visible anywhere) just like the 2charge slam proc, this duration is however not affected by consuming the boosted slam proc itself.
What does this mean? That if you have a 10sec duration proc with 2 charges, and you consume both charges in for example, 3 seconds, you lose the proc because you consumed both charges, but 70849 is still active for 7 seconds more. Checked this with .list auras, it is still among my auras after consuming the boosted proc.
If you get a normal slam proc in the next 7seconds at this point, it will have 2 charges and 10s duration, looking like a t10 proc, because it will be affected by the previous 70849, however sometimes won't have the reduced GCD, likely because the other effect was properly consumed. 
So technically, it's like a fake t10 proc, because it was never supposed to be one, it's modified by the effect of the previous which should no longer be active.
The way the GCD boost 71072 behaves is still a mistery to me, it seems to have 15sec duration (its hard to measure since its also a hidden aura), and what I can't figure is how the it sometimes gets lost in the ether, when it seems to have 15s duration and usually lasts way longer than the effect itself.

TL;DR i think the root of the issue is that the part effects (70849 and 71072) responsible for the enhancing of the slam proc are not consumed with the slam proc and have different durations that don't always line up. I'm also not sure why I have the 4p bonus aura twice (70847) and if it has anything to do with the other auras acting weird (since i suppose its the one that triggers the rest)

**Comment by Dranah** [comment]
Summary (fake t10 proc bug)
- When you get a t10 proc, you activate 46916 (slam!), 70849 (extra charge!) and 71072 (slam gcd effect)
- When you consume both charges, you lose 46916 and 71072, but retain 70849 with whatever is left from its 10sec duration
- If you get a new slam proc within few seconds, before 70849 expires, and if its a t10 one, it reactivates 71072 and all good
- If you get a new slam proc within few seconds, before 70849 expires, and if its regular one, it consumed 70849, giving you a t10 proc but without the gcd boost.
- If you only consumed one charge from the first t10 proc, 71072 will be still active with 1 charge, so you'll have 1 boosted charge on the next one because 71072 also lasts 15sec total.

The other issue is the overwriting below 5sec that needs fixing.
I guess the slam! message stays as-is, although it would have made the entire investigation much more ibv

**Comment by Keader** [comment]
Test now.

**Comment by Dranah** [comment]
Looks good.
No more fake t10 procs, you either get a regular proc with regular gcd, or a t10 proc with increased.
Regular slam! can no longer overwrite the extra slam effect while it has still 2 charges
